### ISSUE TITLE
Bottleneck in file_handle_cache_stats for persistent messages ?

### ISSUE THREAD

**Issue by micoq** [rabbitmq/rabbitmq-server#1664]
`file_handle_cache_stats` and related functions seem to be a bottleneck when using persistent messages or lazy queues.
The slowdown doesn't seem to came from the I/O themselves, the performances are the same on a SSD or on a slow rotating disk.

By using the PerfTest utility, I made 2 tests (1 producer, 1 consumer and 1 queue):

The first test delivers transient messages:
```
./rabbitmq-perf-test-2.1.2/bin/runjava com.rabbitmq.perf.PerfTest -t fanout -e nolazy -x 1 -y 1 -u nolazy -h amqp://admin:admin@localhost -s 1000
id: test-134846-222, time: 1,000s, sent: 12243 msg/s, received: 2386 msg/s, min/median/75th/95th/99th latency: 5617/141054/239399/320958 µs
id: test-134846-222, time: 2,001s, sent: 11695 msg/s, received: 10404 msg/s, min/median/75th/95th/99th latency: 385185/795204/861527/953292 µs
id: test-134846-222, time: 3,001s, sent: 10839 msg/s, received: 13262 msg/s, min/median/75th/95th/99th latency: 655746/849025/893330/986368 µs
id: test-134846-222, time: 4,001s, sent: 14121 msg/s, received: 11741 msg/s, min/median/75th/95th/99th latency: 700232/827872/887429/918279 µs
id: test-134846-222, time: 5,001s, sent: 10840 msg/s, received: 13043 msg/s, min/median/75th/95th/99th latency: 673272/860118/895990/922111 µs
id: test-134846-222, time: 6,001s, sent: 13626 msg/s, received: 11835 msg/s, min/median/75th/95th/99th latency: 690573/815909/874413/910581 µs
id: test-134846-222, time: 7,001s, sent: 10839 msg/s, received: 12641 msg/s, min/median/75th/95th/99th latency: 678320/870186/901713/936544 µs
id: test-134846-222, time: 8,003s, sent: 14093 msg/s, received: 12068 msg/s, min/median/75th/95th/99th latency: 671198/789343/869638/924291 µs
id: test-134846-222, time: 9,003s, sent: 11334 msg/s, received: 12810 msg/s, min/median/75th/95th/99th latency: 674026/859051/899977/927711 µs
id: test-134846-222, time: 10,004s, sent: 13612 msg/s, received: 12116 msg/s, min/median/75th/95th/99th latency: 680650/789207/871676/910717 µs
```
The second test delivers persistent messages:
```
./rabbitmq-perf-test-2.1.2/bin/runjava com.rabbitmq.perf.PerfTest -t fanout -e nolazy -x 1 -y 1 -u nolazy -h amqp://admin:admin@localhost -s 1000 -f persistent
id: test-135107-729, time: 1,001s, sent: 12097 msg/s, received: 1195 msg/s, min/median/75th/95th/99th latency: 9974/178919/303279/385144 µs
id: test-135107-729, time: 2,001s, sent: 7447 msg/s, received: 5596 msg/s, min/median/75th/95th/99th latency: 441788/818719/955883/1183543 µs
id: test-135107-729, time: 3,001s, sent: 4685 msg/s, received: 6514 msg/s, min/median/75th/95th/99th latency: 1217829/1564741/1819645/1885496 µs
id: test-135107-729, time: 4,006s, sent: 4541 msg/s, received: 6788 msg/s, min/median/75th/95th/99th latency: 1338143/1720311/1883147/1978503 µs
id: test-135107-729, time: 5,006s, sent: 10631 msg/s, received: 6080 msg/s, min/median/75th/95th/99th latency: 1369782/1705660/1837127/1923198 µs
id: test-135107-729, time: 6,006s, sent: 4564 msg/s, received: 6675 msg/s, min/median/75th/95th/99th latency: 1316818/1715385/1780282/1876878 µs
id: test-135107-729, time: 7,006s, sent: 5256 msg/s, received: 7398 msg/s, min/median/75th/95th/99th latency: 1265362/1632183/1763616/1844693 µs
id: test-135107-729, time: 8,008s, sent: 9919 msg/s, received: 6226 msg/s, min/median/75th/95th/99th latency: 1303995/1611822/1715002/1848041 µs
id: test-135107-729, time: 9,008s, sent: 4565 msg/s, received: 7122 msg/s, min/median/75th/95th/99th latency: 1267907/1670489/1743200/1792348 µs
id: test-135107-729, time: 10,008s, sent: 7458 msg/s, received: 6905 msg/s, min/median/75th/95th/99th latency: 1293951/1630986/1743476/1861388 µs
```
Note: the results are the same on a lazy queue with both transient or persistent messages.

Transient messages: ~6000 m/s
Persistent messages (or lazy queue): ~12000 m/s

`iotop` and `iostat` doesn't show any activity on the storage (the `*.idx` files are written in `/var/lib/rabbitmq/mnesia/rabbit\@localhost/msg_stores/vhosts/<vhost_id>/queues/<queue_id>` but they are deleted quickly so they only reside in the file system cache)

Using the message store (with `queue_index_embed_msgs_below` set at 0) or set the message size to zero doesn't change anything.

Changing the number of schedulers or IO schedulers on the erlang VM doesn't change anything too.

Finally, I used `eprof` while running the tests and found some statistics functions spend a lot of time:
```
ets:update_counter/3
erlang:monotonic_time/0
erlang:convert_time_unit/3
erts_internal:time_unit/0
file_handle_cache_stats:timer_tc/1
file_handle_cache_stats:update/1  
file_handle_cache_stats:update/2  
```

`file_handle_cache_stats:update/2` seems to be called one on more times per message (1 time if there is no consumer, 3 times with one consumer). This function updates the counter `File handle open attempt` as showed on the management interface. It seems the file handle (of the index file) is checked for each message, is it the expected behavior ?

Here is the complete `eprof` trace:
```
FUNCTION                                                                 CALLS        %      TIME  [uS / CALLS]
--------                                                                 -----  -------      ----  [----------]
rabbit_policy:effective_definition0/2                                        4     0.00         4  [      1.00]
rabbit_queue_consumers:'-count/0-lc$^0/1-0-'/1                               8     0.00         5  [      0.63]
rabbit_policy:name_op/1                                                      4     0.00         5  [      1.25]
rabbit_queue_consumers:'-unacknowledged_message_count/0-lc$^0/1-0-'/1        8     0.00         9  [      1.13]
gen_event:notify/2                                                           4     0.00        10  [      2.50]
rabbit_core_metrics:queue_stats/2                                            4     0.00        10  [      2.50]
rabbit_policy:get/2                                                          8     0.00        12  [      1.50]
prim_file:'-drv_close/1-after$^0/0-0-'/1                                    21     0.00        15  [      0.71]
filename:unix_pathtype/1                                                    13     0.00        15  [      1.15]
prim_file:open_int_setopts/3                                                21     0.00        17  [      0.81]
filename:pathtype/1                                                         13     0.00        17  [      1.31]
rabbit_policy:name0/1                                                        8     0.00        17  [      2.13]
rabbit_mirror_queue_misc:module/1                                            8     0.00        17  [      2.13]
rabbit_policy:effective_definition/1                                         4     0.00        18  [      4.50]
rabbit_policy:name/1                                                         4     0.00        20  [      5.00]
rabbit_priority_queue:set_ram_duration_target/2                              4     0.00        20  [      5.00]
prim_file:'-drv_command_nt/3-after$^0/0-0-'/1                               21     0.00        21  [      1.00]
rabbit_policy:get0/3                                                         8     0.00        25  [      3.13]
erlang:get/0                                                                16     0.00        26  [      1.63]
prim_file:position/2                                                        21     0.00        28  [      1.33]
prim_file:open_mode/1                                                       21     0.00        28  [      1.33]
prim_file:drv_close/1                                                       21     0.00        29  [      1.38]
application_controller:get_env/2                                            42     0.00        29  [      0.69]
prim_file:drv_command_nt/3                                                  21     0.00        31  [      1.48]
rabbit_queue_consumers:all_ch_record/0                                      16     0.00        31  [      1.94]
rabbit_priority_queue:ram_duration/1                                         4     0.00        31  [      7.75]
rabbit_core_metrics:queue_stats/5                                            4     0.00        32  [      8.00]
rabbit_backing_queue:info_keys/0                                             4     0.00        33  [      8.25]
prim_file:drv_open/2                                                        21     0.00        35  [      1.67]
filename:major_os_type/0                                                    13     0.00        36  [      2.77]
lists:reverse/2                                                             42     0.00        38  [      0.90]
prim_file:sync/1                                                            21     0.00        40  [      1.90]
erlang:open_port/2                                                          21     0.00        42  [      2.00]
prim_file:pathname/1                                                        21     0.00        43  [      2.05]
os:type/0                                                                   26     0.00        47  [      1.81]
prim_file:reverse/1                                                         42     0.00        49  [      1.17]
prim_file:'-drv_command/4-after$^0/0-0-'/1                                  91     0.00        49  [      0.54]
rabbit_queue_consumers:'-all_ch_record/0-lc$^0/1-0-'/1                     208     0.00        49  [      0.24]
filename:maybe_remove_dirsep/2                                              13     0.00        54  [      4.15]
maps:to_list/1                                                               7     0.00        56  [      8.00]
maps:fold/3                                                                  7     0.00        57  [      8.14]
prim_file:write/2                                                           21     0.00        61  [      2.90]
erlang:iolist_size/1                                                        42     0.00        62  [      1.48]
rabbit_queue_consumers:count/0                                               8     0.00        65  [      8.13]
gen_event:send/2                                                             4     0.00        70  [     17.50]
prim_file:drv_command/2                                                     91     0.00        72  [      0.79]
prim_file:drv_command/4                                                     91     0.00        72  [      0.79]
rabbit_queue_consumers:unacknowledged_message_count/0                        8     0.00        75  [      9.38]
erlang:port_close/1                                                         21     0.00        75  [      3.57]
prim_file:get_uint64/1                                                      63     0.00        77  [      1.22]
prim_file:drv_command/3                                                     91     0.00        78  [      0.86]
prim_file:open_mode/4                                                      105     0.00        79  [      0.75]
rabbit_memory_monitor:report_ram_duration/2                                  4     0.00        82  [     20.50]
rabbit_mirror_queue_misc:is_mirrored/1                                       8     0.00        84  [     10.50]
prim_file:lseek_position/1                                                  42     0.00        93  [      2.21]
erlang:system_info/1                                                        26     0.00        93  [      3.58]
prim_file:drv_get_response/1                                               112     0.00       104  [      0.93]
ets:lookup/2                                                                42     0.00       104  [      2.48]
prim_file:open_int/4                                                        42     0.00       108  [      2.57]
prim_file:translate_response/2                                             112     0.00       112  [      1.00]
ets:insert/2                                                                12     0.00       112  [      9.33]
file_handle_cache_stats:update/3                                            21     0.00       154  [      7.33]
prim_file:internal_name2native/1                                            21     0.00       162  [      7.71]
prim_file:drv_get_response/2                                               112     0.00       172  [      1.54]
prim_file:truncate/1                                                         7     0.00       173  [     24.71]
application:get_env/2                                                       42     0.00       178  [      4.24]
prim_file:get_uint32/1                                                     126     0.00       185  [      1.47]
erlang:port_command/2                                                      112     0.00       197  [      1.76]
prim_file:open/2                                                            21     0.00       199  [      9.48]
rabbit_core_metrics:gen_server2_stats/2                                      4     0.00       269  [     67.25]
erts_internal:open_port/2                                                   21     0.00       594  [     28.29]
maps:'-fold/3-fun-0-'/3                                                    217     0.00       680  [      3.13]
erts_internal:port_close/1                                                  21     0.00       696  [     33.14]
filename:join1/4                                                          1638     0.00       737  [      0.45]
erlang:bump_reductions/1                                                   112     0.00       881  [      7.87]
prim_file:close/1                                                           21     0.01      1591  [     75.76]
filename:join/2                                                             13     0.08     14526  [   1117.38]
rabbit_queue_consumers:update_use/2                                     199951     0.57    109515  [      0.55]
erts_internal:port_command/3                                               112     0.70    134025  [   1196.65]
rabbit_queue_consumers:inactive/1                                       199959     0.84    162028  [      0.81]
file_handle_cache_stats:timer_tc/1                                      200063     0.86    164760  [      0.82]
rabbit_queue_consumers:deliver/4                                        199951     1.35    259299  [      1.30]
rabbit_queue_consumers:deliver/3                                        199951     1.42    273697  [      1.37]
erts_internal:time_unit/0                                               207995     1.45    278847  [      1.34]
erlang:convert_time_unit/3                                              207995     1.50    288460  [      1.39]
rabbit_priority_queue:is_duplicate/2                                    199951     2.12    408026  [      2.04]
rabbit_priority_queue:info/2                                            399962     3.91    752943  [      1.88]
rabbit_priority_queue:len/1                                             599861     4.19    805252  [      1.34]
rabbit_priority_queue:drain_confirmed/1                                 199959     4.39    844986  [      4.23]
rabbit_priority_queue:publish/6                                         199951     4.50    864858  [      4.33]
ets:update_counter/3                                                    600119     5.02    965584  [      1.61]
erlang:monotonic_time/0                                                 400126     5.57   1071299  [      2.68]
rabbit_priority_queue:needs_timeout/1                                   199958     7.99   1537533  [      7.69]
file_handle_cache_stats:update/1                                        199972     9.03   1737185  [      8.69]
erlang:setelement/3                                                     799819    20.79   3999071  [      5.00]
file_handle_cache_stats:update/2                                        200042    23.68   4554546  [     22.77]
---------------------------------------------------------------------  -------  -------  --------  [----------]
Total:                                                                 5420019  100.00%  19236136  [      3.55]
```

RabbitMQ version: 3.7.2
Erlang version: 20.2
OS: RHEL 7.4
Kernel: 3.10.0-693.11.6.el7.x86_64
Client:
* Perf Test version: 2.1.2
* Java: 1.8.0
* RabbitMQ AMQP Client version: 5.2.0

**Comment by michaelklishin** [comment]
You can see that there's quite a number of functions that may be major contributors. I suggest this is moved to the mailing list for now.

**Comment by michaelklishin** [comment]
Note that we [explicitly mention](http://www.rabbitmq.com/queues.html#runtime-characteristics) in the docs that a single queue will become a bottleneck regardless of what function ends up being the most significant contributor.

**Comment by micoq** [comment]
@michaelklishin I recompiled RabbitMQ from the latest sources and removed manually the calls to the `file_handle_cache_stats:*` functions and the behavior remain the same so the `eprof` results are probably not as relevant as I thought.

For your second comment, I know a single queue is backed by a single Erlang process and I already use several queues to distribute the load. I tried to understand why a single lazy queue (or with persistent messages) is twice as slow as a ram-only queue with transient messages (and with no apparent storage activity).
